File: avot-core/adapters/signal_ledger.py
from typing import Dict, Any, List, Optional
from pathlib import Path
from datetime import datetime
import yaml
import uuid


class SignalLedgerError(Exception):
    """Raised when the signal ledger cannot be read or written safely."""
    pass


class SignalLedger:
    """
    Read / append-only adapter for the Signal Ledger.

    This adapter:
    - reads the ledger as truth
    - appends new signals
    - never mutates or deletes existing signals
    - performs no interpretation
    """

    def __init__(self, ledger_path: str):
        self.ledger_path = Path(ledger_path)

        if not self.ledger_path.exists():
            raise SignalLedgerError(
                f"Signal ledger not found at {self.ledger_path}"
            )

    # --- Internal helpers ---

    def _load(self) -> Dict[str, Any]:
        try:
            with self.ledger_path.open("r", encoding="utf-8") as f:
                return yaml.safe_load(f) or {}
        except Exception as e:
            raise SignalLedgerError(f"Failed to read signal ledger: {e}")

    def _write(self, data: Dict[str, Any]) -> None:
        try:
            with self.ledger_path.open("w", encoding="utf-8") as f:
                yaml.safe_dump(data, f, sort_keys=False)
        except Exception as e:
            raise SignalLedgerError(f"Failed to write signal ledger: {e}")

    # --- Public API ---

    def read_signals(self) -> List[Dict[str, Any]]:
        ledger = self._load()
        return ledger.get("signal_ledger", {}).get("signals", [])

    def append_signal(
        self,
        avot_id: str,
        signal_type: str,
        description: str,
        severity: str = "low",
        context: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Append a new signal emitted by an AVOT.

        This method:
        - generates a unique signal_id
        - timestamps the observation
        - appends without altering existing entries
        """

        ledger = self._load()

        if "signal_ledger" not in ledger:
            raise SignalLedgerError("Invalid signal ledger structure")

        signal = {
            "signal_id": f"SIG-{uuid.uuid4()}",
            "avot_id": avot_id,
            "signal_type": signal_type,
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "severity": severity,
            "description": description,
            "context": context or {},
            "metadata": metadata or {},
        }

        ledger["signal_ledger"].setdefault("signals", []).append(signal)
        self._write(ledger)

        return signal

File: avot-core/adapters/signalLedger.ts
import fs from "fs";
import yaml from "js-yaml";
import { v4 as uuidv4 } from "uuid";

export class SignalLedgerError extends Error {}

export class SignalLedger {
  private ledgerPath: string;

  constructor(ledgerPath: string) {
    this.ledgerPath = ledgerPath;

    if (!fs.existsSync(this.ledgerPath)) {
      throw new SignalLedgerError(
        `Signal ledger not found at ${this.ledgerPath}`
      );
    }
  }

  private load(): any {
    try {
      const raw = fs.readFileSync(this.ledgerPath, "utf-8");
      return yaml.load(raw) || {};
    } catch (err) {
      throw new SignalLedgerError(
        `Failed to read signal ledger: ${(err as Error).message}`
      );
    }
  }

  private write(data: any): void {
    try {
      fs.writeFileSync(
        this.ledgerPath,
        yaml.dump(data, { sortKeys: false }),
        "utf-8"
      );
    } catch (err) {
      throw new SignalLedgerError(
        `Failed to write signal ledger: ${(err as Error).message}`
      );
    }
  }

  readSignals(): any[] {
    const ledger = this.load();
    return ledger?.signal_ledger?.signals ?? [];
  }

  appendSignal(
    avotId: string,
    signalType: string,
    description: string,
    severity: string = "low",
    context: Record<string, unknown> = {},
    metadata: Record<string, unknown> = {}
  ): any {
    const ledger = this.load();

    if (!ledger.signal_ledger) {
      throw new SignalLedgerError("Invalid signal ledger structure");
    }

    const signal = {
      signal_id: `SIG-${uuidv4()}`,
      avot_id: avotId,
      signal_type: signalType,
      timestamp: new Date().toISOString(),
      severity,
      description,
      context,
      metadata,
    };

    ledger.signal_ledger.signals.push(signal);
    this.write(ledger);

    return signal;
  }
}
